be/src/runtime/sender_queue.h

```diff
@@ -0,0 +1,237 @@
+    // key of second level is request sequence
+    phmap::flat_hash_map<int, phmap::flat_hash_map<int64_t, ChunkQueue>> _buffered_chunk_queues;
+};
+
+// PipelineSenderQueue will be called in the pipeline execution threads.
+// In order to avoid thread blocking caused by lock competition,
+// we try to use lock-free structures in the relevant interface as much as possible.
+// It should be noted that some atomic variables are updated at the same time under the protection of lock,
+// which may not be completely consistent when reading without lock, but will eventually be consistent.
+// This won't affect the correctness in our usage scenario.
+class DataStreamRecvr::PipelineSenderQueue : public DataStreamRecvr::SenderQueue {
+public:
+    PipelineSenderQueue(DataStreamRecvr* parent_recvr, int num_senders, int degree_of_parallism);
+    ~PipelineSenderQueue() override = default;
+
+    Status get_chunk(vectorized::Chunk** chunk, const int32_t driver_sequence = -1) override;
+
+    bool has_chunk() override;
+
+    bool try_get_chunk(vectorized::Chunk** chunk) override;
+
+    Status add_chunks(const PTransmitChunkParams& request, ::google::protobuf::Closure** done) override;
+
+    Status add_chunks_and_keep_order(const PTransmitChunkParams& request, ::google::protobuf::Closure** done) override;
+
+    void decrement_senders(int be_number) override;
+
+    void cancel() override;
+
+    void close() override;
+
+    void short_circuit(const int32_t driver_sequence);
+
+    bool has_output(const int32_t driver_sequence);
+
+    bool is_finished() const;
+
+private:
+    void clean_buffer_queues();
+
+    struct ChunkItem {
+        int64_t chunk_bytes = 0;
+        // Invalid if SenderQueue::_is_pipeline_level_shuffle is false
+        int32_t driver_sequence = -1;
+        ChunkUniquePtr chunk_ptr;
+        // When the memory of the ChunkQueue exceeds the limit,
+        // we have to hold closure of the request, so as not to let the sender continue to send data.
+        // A Request may have multiple Chunks, so only when the last Chunk of the Request is consumed,
+        // the callback is closed- >run() Let the sender continue to send data
+        google::protobuf::Closure* closure = nullptr;
+    };
+
+    class ChunkQueue {
+    public:
+        ChunkQueue() {}
+        ~ChunkQueue() {}
+
+        ChunkQueue(const ChunkQueue&) = delete;
+        ChunkQueue& operator=(const ChunkQueue&) = delete;
+
+        ChunkQueue(ChunkQueue&& rhs) {
+            _chunks.swap(rhs._chunks);
+            _size.store(rhs._size);
+        }
+
+        ChunkQueue& operator=(ChunkQueue&& rhs) {
+            _chunks.swap(rhs._chunks);
+            _size.store(rhs._size);
+            return *this;
+        }
+
+        inline bool empty() const { return _size.load() == 0; }
+
+        inline bool size() const { return _size.load(); }
+
+        void enqueue(ChunkItem&& item) {
+            _chunks.enqueue(std::move(item));
```

Comment:
Why not use `ConcurrentQueue::size()`, cause `_size` may be unconsistent with `_chunks.size()`?